dom.go

```diff
@@ -1,92 +1,164 @@
+// Core implements the Context method of the Component interface, and is the
+// core/central struct which all Component implementations should embed.
+type Core struct {
+	prevRender *HTML
 }
 
-// RenderAsBody renders the given component as the body of the page, replacing
-// whatever existing content in the page body there may be.
-func RenderAsBody(comp Component) {
-	doc := js.Global.Get("document")
-	body := doc.Call("createElement", "body")
-	Render(comp, body)
-	if doc.Get("readyState").String() == "loading" {
-		doc.Call("addEventListener", "DOMContentLoaded", func() { // avoid duplicate body
-			doc.Set("body", body)
-		})
-		return
-	}
-	doc.Set("body", body)
-}
+// Context implements the Component interface.
+func (c *Core) Context() *Core { return c }
 
-type textComponent struct {
-	text string
-	node *js.Object
-}
+// Component represents a single visual component within an application. To
+// define a new component simply implement the Render method and embed the Core
+// struct:
+//
+// 	type MyComponent struct {
+// 		vecty.Core
+// 		... additional component fields (state or properties) ...
+// 	}
+//
+// 	func (c *MyComponent) Render() *vecty.HTML {
+// 		... rendering ...
+// 	}
+//
+type Component interface {
+	// Render is responsible for building HTML which represents the component.
+	Render() *HTML
 
-// Apply implements the Markup interface.
-func (s *textComponent) Apply(element *Element) {
-	element.Children = append(element.Children, s)
+	// Context returns the components context, which is used internally by
+	// Vecty in order to store the previous component render for diffing.
+	Context() *Core
 }
 
-func (s *textComponent) Reconcile(oldComp Component) {
-	if oldText, ok := oldComp.(*textComponent); ok {
-		s.node = oldText.node
-		if oldText.text != s.text {
-			s.node.Set("nodeValue", s.text)
-		}
-		return
-	}
+// ComponentOrHTML represents one of:
+//
+//  Component
+//  *HTML
+//
+// If the underlying value is not one of these types, the code handling the
+// value is expected to panic.
+type ComponentOrHTML interface{}
 
-	s.node = js.Global.Get("document").Call("createTextNode", s.text)
+// Restorer is an optional interface that Component's can implement in order to
+// restore state during component reconciliation and also to short-circuit
+// the reconciliation of a Component's body.
+type Restorer interface {
+	// Restore is called when the component should restore itself against a
+	// previous instance of a component. The previous component may be nil or
+	// of a different type than this Restorer itself, thus a type assertion
+	// should be used.
+	//
+	// If skip = true is returned, restoration of this component's body is
+	// skipped. That is, the component is not rerendered. If the component can
+	// prove when Restore is called that the HTML rendered by Component.Render
+	// would not change, true should be returned.
+	Restore(prev Component) (skip bool)
 }
```

Comment:
- `Reconcile` is now named `Restore`. In terms of English, `Reconcile` is quite the loaded word to spell and `Restore` has the same meaning. I think this will be a win for people not used to React wording, but a small loss for those who are used to React wording. Simple naming == better naming, so we'll go with this.
- How to short-circuit component restoration is now documented :) Fixes #27